### PR TITLE
Explicit check of chainerx arrays on fallback functions

### PR DESCRIPTION
**Pull request by emcastillo**
We don't allow numpy or cupy arrays to reach fallback functions on chainer x, and we also forbid advanced indexing of chainerx arrays using numpy or cupy arrays.

The behavior is according to the following code:
```
x = chx.array([[0., 1., 2.], [3., 4., 5.]],dtype=chx.float32,device='cuda:0')
xnp = np.array([[0., 1., 2.], [3., 4., 5.]])
x_i = (chx.array([1],device='cuda:0'),)
x_inp = np.array([1])
x_icp = cp.array([1])
x_tup = ([1],)
print(chx.fix(x)) #ALLOWED
print(chx.fix(xnp)) #TYPE ERROR
print('CX ',x[x_i]) #ALLOWED
print('NP ',x[x_inp])#TYPE ERROR
print('CP ',x[x_icp])#TYPE ERROR
print('TU ',x[x_tup])#ALLOWED
```

### CODE REVIEW

File: chainerx/_fallback_workarounds.py [hunk 1]
```diff
@@ -62,6 +62,13 @@ def _from_chx(array, check_backprop=True):
     # Objects with other types are kept intact.
     # Returns a pair: (xp, cupy device or dummy context, numpy/cupy.ndarray).
     if not isinstance(array, chainerx.ndarray):
+        if (isinstance(array, numpy.ndarray) or
+                cupy and isinstance(array, cupy.ndarray)):
```
Comment (niboshi) [inline chainerx/_fallback_workarounds.py]: A nitpick, but it better to bring `or` to the front of the next line.
(See [PEP-8](https://www.python.org/dev/peps/pep-0008/#should-a-line-break-before-or-after-a-binary-operator).)
Also, as this expression mixes `or` and `and`, it's better to put parentheses around the `and` expression.

Comment (emcastillo) [reply chainerx/_fallback_workarounds.py]: Changed! Thanks :D

**Comment by niboshi** [comment]
Thanks for the fix.
Jenkins, test this please

**Comment by niboshi** [comment]
Jenkins, test this please

**Comment by niboshi** [comment]
LGTM

**Pull request merged**